### PR TITLE
Add `stats_bucket` / `extended_stats_bucket` pipeline aggs

### PR DESCRIPTION
**Pull request by polyfractal**
These are the complements to the stats/extended_stats metric aggregations, and can be used
to calculate a variety of statistics over buckets

Pretty straight forward addition, but a few notes:
- These extend `metrics.stats.InternalStats` & `metrics.stats.extended.ExtendedInternalStats` to reduce copy/paste of basically the same code
- The pipeline aggs register the stream of their specific internal representation (see `StatsBucketPipelineAggregator.registerStreams()` for example).  I thought this was cleaner than cluttering up the SearchModule registration list with 2x registrations for each, but happy to change if this is too indirect
- Provides two interfaces (`StatsBucket` & `ExtendedStatsBucket`) which are basically just identical subclasses of the metric versions.  Thought it would still be good to have these for the end-user.

/cc @colings86 if you have some spare time, no rush though.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/pipeline/bucketmetrics/stats/InternalStatsBucket.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.aggregations.support.format.ValueFormatter;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public class InternalStatsBucket extends InternalStats implements StatsBucket {
+
+    public final static Type TYPE = new Type("internal_stats_bucket");
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/bucketmetrics/stats/InternalStatsBucket.java]: I would use the same type name as the StatsPipelineAggregator here so it's easy to see they relate. That's what we do on other aggregations. The contexts in which these are used and deserialised are different so the names would never clash

File: core/src/main/java/org/elasticsearch/search/aggregations/pipeline/bucketmetrics/stats/extended/InternalExtendedStatsBucket.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.aggregations.support.format.ValueFormatter;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public class InternalExtendedStatsBucket extends InternalExtendedStats implements ExtendedStatsBucket {
+
+    public final static Type TYPE = new Type("internal_extended_stats_bucket");
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/bucketmetrics/stats/extended/InternalExtendedStatsBucket.java]: I would use the same type name as the StatsPipelineAggregator here so it's easy to see they relate. That's what we do on other aggregations. The contexts in which these are used and deserialised are different so the names would never clash

**Comment by colings86** [comment]
Left a couple of comments but I think this is looking good.

**Comment by polyfractal** [comment]
Comments addressed!

**Comment by colings86** [comment]
LGTM

**Pull request merged**